internal/handlers/pg/pgdb/query.go

```diff
@@ -246,20 +247,42 @@ func prepareWhereClause(sqlFilters *types.Document) (string, []any) {
 	var p Placeholder
 
 	for k, v := range sqlFilters.Map() {
-		switch k {
-		case "_id":
-			switch v := v.(type) {
-			case string:
-				filters = append(filters, fmt.Sprintf(`((_jsonb->'_id')::jsonb = %s)`, p.Next()))
-				args = append(args, string(must.NotFail(pjson.MarshalSingleValue(v))))
+		if len(k) != 0 && (k[0] == '$' || types.NewPathFromString(k).Len() > 1) {
+			// TODO $eq and $ne https://github.com/FerretDB/FerretDB/issues/1840
+			// TODO dot notation https://github.com/FerretDB/FerretDB/issues/1841
+			continue
+		}
 
-			case types.ObjectID:
+		// don't iterate through array for _id keys to simplify the query
+		if k == "_id" {
+			switch v := v.(type) {
+			case string, types.ObjectID:
 				filters = append(filters, fmt.Sprintf(`((_jsonb->'_id')::jsonb = %s)`, p.Next()))
 				args = append(args, string(must.NotFail(pjson.MarshalSingleValue(v))))
 			}
-		default:
+
 			continue
 		}
+
+		switch v := v.(type) {
+		case *types.Document, *types.Array, types.Binary, bool, time.Time, types.NullType, types.Regex, types.Timestamp:
+			continue
+
+		case float64, string, types.ObjectID, int32, int64:
+			sql := fmt.Sprintf(
+				`((_jsonb->%[1]s)::jsonb = %[2]s)`+ // Select if value under the key k is equal to value v.
+					// If it's not, but the value under the key k is an array - select if it contains the value equal to v.
+					` OR (jsonb_typeof(_jsonb->%[1]s) = 'array' AND`+
```

Comment:
Do we have tests for array? We should have one that matches and one that doesn't match this condition. Also empty array should be tested :hugs:

Reply:
We already test for that. For example when we run int64 or int32 tests they'll search for top level `42` values but also for arrays with `42`:

```
=== RUN   TestQueryComparisonCompatImplicit/Int64/TestQueryComparisonCompatImplicit_ArrayInt32s
=== CONT  TestQueryComparisonCompatImplicit/Int64/TestQueryComparisonCompatImplicit_ArrayInt32s
    query_comparison_compat_test.go:181: Compat (expected) IDs: [array-int32-one array-int32-three array-int32-two]
    query_comparison_compat_test.go:181: Target (actual)   IDs: [array-int32-one array-int32-three array-int32-two]
        --- PASS: TestQueryComparisonCompatImplicit/Int64/TestQueryComparisonCompatImplicit_ArrayInt32s (0.01s)
```

Reply:
Thank :+1: